modules/imgproc/src/connectedcomponents.cpp

```diff
@@ -1737,14 +1755,40 @@ int cv::connectedComponentsWithStats(InputArray _img, OutputArray _labels, Outpu
     _labels.create(img.size(), CV_MAT_DEPTH(ltype));
     cv::Mat labels = _labels.getMat();
     connectedcomponents::CCStatsOp sop(statsv, centroids);
-    if (ltype == CV_16U){
-        return connectedComponents_sub1(img, labels, connectivity, ccltype, sop);
+    int nbComponents;
+    if (ltype == CV_16U) {
+        nbComponents = connectedComponents_sub1(img, labels, connectivity, ccltype&(~CCL_FORE_BACK_GROUND), sop);
     }
-    else if (ltype == CV_32S){
-        return connectedComponents_sub1(img, labels, connectivity, ccltype, sop);
+    else if (ltype == CV_32S) {
+        nbComponents = connectedComponents_sub1(img, labels, connectivity, ccltype&(~CCL_FORE_BACK_GROUND), sop);
     }
-    else{
+    else {
         CV_Error(CV_StsUnsupportedFormat, "the type of labels must be 16u or 32s");
         return 0;
     }
+    if (!(ccltype & CCL_FORE_BACK_GROUND))
+        return nbComponents;
+    bitwise_not(img, img);
```

Comment:
I updated my repo so now this fix is in my code.  But exception is still here with this code : 
```
         Mat img = imread("f:/lib/opencv/samples/data/box.png", IMREAD_GRAYSCALE),imgThresh, imgThreshNot;
        Mat labelsBackground;
        Mat statsvBackground, centroidBackground;

        double thresh = threshold(img, imgThresh, 50, 255, THRESH_OTSU);
        bitwise_not(imgThresh, imgThreshNot);
        int nbLabelsBackGround = connectedComponentsWithStats(imgThreshNot, labelsBackground, statsvBackground, centroidBackground, 4, CV_8U, CCL_WU);

```
I thnik problem is at  [this line ](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/connectedcomponents.cpp#L207) LabelT is uchar and [at line](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/connectedcomponents.cpp#L207) there is no check lunique is equal to 255. Then[ sop dimension](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/connectedcomponents.cpp#L207) is wrong  and exception. Bug should exist for CV_16U and CV_32U . May be some CV_ASSERT should be inserted